Make animation and router support simulation, and use Spring for Cupertino

dkwingsmt opened the pull request:
This PR makes `TransitionRoute` support driving the animation with `Simulation`. This is needed for Cupertino widgets, since iOS use "spring simulations" for a majority of their animations.

This PR also applies the standard spring animation to `CupertinoDialogRoute` (alert dialogs) and `CupertinoModalPopupRoute` (action sheets). (This PR does not yet support customizing the spring parameters or conveniently using the springs for custom routes, which are left for future PRs.)

### Comparison

I tried to create a comparison video for action sheets, however the difference is far less noticeable than I expected. (All clips are precisely aligned at the moment the pointer is lifted.) I guess the original curve _is_ good enough. Nevertheless, the spring simulation is the correct one and we should support it.

Edit: [The comment](https://github.com/flutter/flutter/pull/155575#issuecomment-2489303282) below also mentioned that supporting spring animation will improve the fidelity when the animation is caused by a user gesture. I assume this requires initial speed, which is not supported by this PR but we can add it in the future.

https://github.com/user-attachments/assets/06d2f684-ad1c-4a4d-8663-a561895f45e9

Also, Flutter's response seems to be always a moment slower than SwiftUI, possibly because Flutter requiring two frames to start the animation (one frame to add the transition widget, one frame for the animation to actually progress.) We probably want to solve it in the future.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

dkwingsmt committed:
Works

dkwingsmt committed:
Remove AC.duration (seems safe)

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into spring-route

dkwingsmt committed:
_createIosSpring

dkwingsmt committed:
Better docs for _createIosSpring

dkwingsmt committed:
debugIsDisposed and remove the need to override AC

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into spring-route

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into spring-route

dkwingsmt committed:
Remove _createIosSpring

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into spring-route

dkwingsmt committed:
Fix compile and split animateBackWith

dkwingsmt committed:
Fix doc

dkwingsmt committed:
Better doc and better way to decide simulation usage

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into spring-route

dkwingsmt committed:
Add test

dkwingsmt committed:
Add tests. But a test doesn't work

dkwingsmt committed:
Test works

dkwingsmt committed:
Test works more

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into spring-route

dkwingsmt committed:
Fix doc

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into spring-route

dkwingsmt committed:
Great action sheet delay

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into spring-route

dkwingsmt committed:
Remove delaying

Piinks commented:
(PR Triage): Flyby comment, there are a bunch of failing tests here. 🙃

MitchellGoodwin commented:
> I guess the original curve is good enough.

Flyby comment. We've done a good job of matching the curves for most transitions in Cupertino when they are predictable. It becomes fuzzy when a transition is caused by a user gesture, like swiping, or dragging and releasing a page. Spring simulations will really shine in those situations.

chunhtai commented on packages/flutter/lib/src/cupertino/route.dart:
```diff
@@ -1433,33 +1426,29 @@ class CupertinoDialogRoute<T> extends RawDialogRoute<T> {
       return super.buildTransitions(context, animation, secondaryAnimation, child);
     }
 
-    if (_fadeAnimation?.parent != animation) {
```
Why do we not applying the curve now?

dkwingsmt replied:
The curve is now driven by the simulation. `Curves.easeInOut` is basically an approximation of the `SpringSimulation`.

chunhtai commented on packages/flutter/lib/src/cupertino/route.dart:
```diff
@@ -1433,33 +1426,29 @@ class CupertinoDialogRoute<T> extends RawDialogRoute<T> {
-        opacity: fadeAnimation,
-        child: super.buildTransitions(context, animation, secondaryAnimation, child),
+        opacity: animation,
+        child: child,
```
why no super?

dkwingsmt replied:
The super's buildTransition is:
```dart
  Widget buildTransitions(BuildContext context, Animation<double> animation, Animation<double> secondaryAnimation, Widget child) {
    if (_transitionBuilder == null) {
      // Some default transition.
      return FadeTransition(
        opacity: animation,
        child: child,
      );
    }
    return _transitionBuilder(context, animation, secondaryAnimation, child);
  }
```
It's some default fading transition, and should not be applied on top of what we have here for cupertino route.

dkwingsmt replied:
(In fact I kind of think the 
```dart
    if (transitionBuilder != null) {
      return super.buildTransitions(context, animation, secondaryAnimation, child);
    }
```
part should not use super's buildTransitions either, but it's not the focus of this PR.)

chunhtai replied:
further looking at the code, is the intention that the CupertinoDialogRoute.transitionbuilder will override all transition in this class?

dkwingsmt replied:
> further looking at the code, is the intention that the CupertinoDialogRoute.transitionbuilder will override all transition in this class?

Yeah, that's how I comprehend the code before the PR. The case where `transitionBuilder` has been processed at the beginning of the method.

chunhtai replied:
sg

chunhtai commented on packages/flutter/lib/src/widgets/routes.dart:
```diff
@@ -213,17 +213,32 @@ abstract class TransitionRoute<T> extends OverlayRoute<T> implements PredictiveB
+  bool debugIsDisposed() {
+    bool disposed = false;
+    assert(() {
+      disposed = _transitionCompleter.isCompleted;
```
this seems to be a weird way to determine dispose. Should this be in the base route class and based on the value _disposeCompleter.isCompleted ?

dkwingsmt replied:
This is simply a refactor, see `createAnimationController` before the PR (the left column) and other methods. I refactored this because `_transitionCompleter` isn't public and I wanted this extra check in my overridden `createSimulation`.

chunhtai replied:
I am fine with having this extra public method, but my concern is that disposed should be mapped to _disposeCompleter not _transitionCompleter. If you really meant to use transitionCompleter, I think this method needs a better name

dkwingsmt replied:
The reason why I chose this name mostly comes from its assert message:
```dart
assert(!_transitionCompleter.isCompleted, 'Cannot reuse a $runtimeType after disposing it.');
```
I'm happy to accept other names, but then I think this assertion message would be wrong.

Considering the completer's Future has the following comment:
```dart
  /// This future completes only once the transition itself has finished, after
  /// the overlay entries have been removed from the navigator's overlay.
  ///
  /// This future completes once the animation has been dismissed. That will be
  /// after [popped], because [popped] typically completes before the animation
  /// even starts, as soon as the route is popped.
```

What do you think of  the name `debugTransitionFinished`? (We might still want to change the assert message though...)

chunhtai replied:
the debugTransitionFinished sounds good to me

chunhtai commented on packages/flutter/lib/src/widgets/routes.dart:
```diff
@@ -236,11 +251,59 @@ abstract class TransitionRoute<T> extends OverlayRoute<T> implements PredictiveB
+  /// See also:
+  ///
+  ///  * [createSimulation], which creates this value.
+  Simulation? get simulation => _simulation;
```
Seems a bit weird that we expose this property. Can we avoid this?

dkwingsmt replied:
We can, but I don't find it especially weird. If a route is driven by a simulation, then it's very reasonable for query the current simulation in progress. In fact, some cases might even need to the simulation object to access its status such as `isDone` for when the animation is considered completed (since it might have a very long tail) or `dx` to smoothly connect simulations.

I can remove this parameter until we really need it, if you insist. What do you think?

chunhtai replied:
I am worry about exposing transient state, this may be error prone if developer calls this at the wrong time. If this is useful in some callback, we should consider it as a pass in parameter to those call back. I vote for removing this now and expose it in the future once we have more use cases.

chunhtai reviewed (commented):
The change and API mostly looks good, left some questions.

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into spring-route

dkwingsmt committed:
Doc

chunhtai reviewed (commented):
left some questions

dkwingsmt committed:
Remove .simulation

dkwingsmt committed:
Tolerance

dkwingsmt committed:
Adjust tests

dkwingsmt committed:
snapToEnd

dkwingsmt committed:
Revert test

dkwingsmt committed:
Better doc for tolerance

dkwingsmt committed:
Extra test

dkwingsmt committed:
Lengthen animation

dkwingsmt commented:
While fixing the bugs I made some nontrivial changes:
- Added `_kStandardTolerance` so that the animation duration is calculated in a similar way as SwiftUI.
- Added parameter `bool snapToEnd` to `SpringSimulation` so that the position is exactly `end` at the end of transition. A corresponding test is also added.
(Also I'm aware that I haven't fix the method name `debugIsDisposed` since I asked for your opinion.)

chunhtai reviewed (commented):
LGTM

dkwingsmt committed:
debugTransitionCompleted

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into spring-route

orestesgaolin commented:
It's not specifically a breaking change but I think it would be super nice to announce to the widget package maintainers that new method is now added to `AnimationController`: `animateBackWith`

dkwingsmt commented:
> It's not specifically a breaking change but I think it would be super nice to announce to the widget package maintainers that new method is now added to `AnimationController`: `animateBackWith`

I see that it broke the packages that implement the API. My sincere apologies for the inconvenience. I know how bad hidden breakages make maintainers feel.

However, I wonder how we should announce "breakages" like this one. Technically every changes are breaking, whether it's fixing a behavioral bug (someone might be relying on it) or adding a new property (which breaks implementations like in your case.) Flutter used to be much more strict on the criteria of breakages, but it didn't scale well, and now only those that affects Flutter's unit tests and customers tests in the flutter/test are considered breaking. In the end, we need to find a fine balance on announcing actually informative breakages while not bloating the patchnotes to too long.

Considering a lot of PRs will add properties to a public class, I don't think listing all of them is a good idea, although honestly I haven't really counted how many. I'll bring this to the internal discussion for viability. Thanks!
